### ISSUE TITLE
💡 [FEATURE REQUEST] Improved Default Zoom for all Kepler Maps

### ISSUE THREAD

**Issue by alexbfree** [hestiaAI/hestialabs-experiences#1093]
Paul pointed out that on maps (e.g. Uber, Google locations) the default view is scoped to show the whole world.

Instead, the map should zoom to something closer to the maximal extent of the points in the data.

It should take whichever is the smaller area of :

- that maximal area
- 200km around the top heard wifi access point location, if we have it

**Comment by alexbfree** [comment]
Paul has clarified the centre point should be:

- Top heard wifi signal average => There is a tab that has wifi signals, for which we show the average location. That location.

Apparently this is different from average location, and can get around issues with someone who moved city. Will need to dig into code to understand how it works.

**Comment by pdehaye** [comment]
I meant my comment for Gmaps.

We didn't use Uber driver when doing the coaching session with client yesterday.

The ticket is formulated at the top with a different scope than what I intended:
- also Uber (so my suggestion not applicable, we don't have this data)
- all maps while some Kepler have to do with points, some with trips, and some with both (I meant for just the first tab of Gmaps which is where we experienced the problems, which might have been present elsewhere)

**Comment by alexbfree** [comment]
I think you are saying:

1. "narrow the scope of this ticket, we only need to change the default zoom for google maps first tab" (plus potentially some others where we have default zoom issues, as yet not identified)

and

2. "for the "minimum of 200m from top heard wifi signal and max extent" rule, that can only be used in the case of google maps data where we have wifi data" -> so presumably that implies we just use maximal datapoint extents in any other case.

have i understood correctly?

If so, please provide additional detail how I can identify what other tabs 1. applies to.

**Comment by pdehaye** [comment]
In the current situation wait for a bug report?

The ideal would be that we could quickly identify all the experience tabs relying on kepler, then those that display points. (That ideal should not be pursued on its own. Many parallel tactical steps go in this direction, such as optimizing Uber driver for digipower live.)

Which makes me think: my Uber passenger is going to be subject to the same problem (I do airdrop around the world), and we should have my own data as a sample dataset and now I am tempted to out this in scope for the *next* digipower live (with home address removed).

**Comment by alexbfree** [comment]
> In the current situation wait for a bug report?

No, not waiting for anything now. I can action this.

It's next up after #1092 

Where you can help is to highlight places where you see issues with default zoom, as you notice them - just add them to this ticket (experience+tab+which dataset). Will be good to have multiple test scenarios.

**Comment by alexbfree** [comment]
BTW presumably since this is now focused scope on google maps that makes it lower priority than the Uber-centric stuff within #1087 that is needed for Thursday?

**Comment by alexbfree** [comment]
Lowering priority as reasoned above, and removing from kepler/uber working set.

**Comment by alexbfree** [comment]
This is also on hold pending resolution of https://github.com/hestiaAI/hestialabs-experiences/issues/1104

**Comment by alexbfree** [comment]
Given #1104 has no way forward, we have to find a work around for developing this (such as use of a local/test CDN URL, or using the Kepler standalone project).

**Comment by alexbfree** [comment]
Unassigned myself in case I don't get to this before going on leave and it needs advancing - but if it still here next week I will pick it up again.

**Comment by alexbfree** [comment]
@pdehaye https://github.com/digipower-academy/private/issues/22 suggests this feature request should be reprioritised. 

could you clarify whether this is higher priority than directus work and how far you want us to go (e.g. do we do it for all experiences, do we do a fixed area (e.g. Europe) as suggested in linked ticket, or do we do it per the data (as described in this ticket above).

**Comment by pdehaye** [comment]
It's a higher priority for you to do this than Directus work.

**Comment by alexbfree** [comment]
Thanks, and any comment on scope please?

**Comment by alexbfree** [comment]
I propose that the best sensible default for all maps might be this general algorithm:

```
1. Scan all data points and find max X, max Y, min X, min Y
2. Calculate Xspan (maxX-minX), Yspan (maxY-maxY)
3. Calculate midpointX ((maxX+minX)/2) and midpointY ((maxY+minY)/2)
4. on the premise we would like all available data to take up about 7/8 of the available window, leaving a small border around so data is not right at the edge, calculate xStep (=Xspan/6) and yStep(=Yspan/6)
5. Set view window coordinates to:

X goes from [midpointX-(4*xStep)] to [midpointX+(4*xStep)]
Y goes from [midpointY-(4*yStep)] to [midpointX+(4*yStep)]
```

I think this should give us a decent view for all data extents, with a view window just outside the extent of the available data.

**Comment by alexbfree** [comment]
I don't think the above algorithm is viable; the [`updateMap` command for KeplerGL](https://docs.kepler.gl/docs/api-reference/actions/actions#updatemap) doesn't allow you to specify min & max viewport extents, instead it needs a centre, and a zoom level.

The zoom levels are a small integer (e.g. 9). These don't seem to be documented in KeplerGL docs but my working assumption is they are the same as [Mapbox Zoom levels](see https://docs.mapbox.com/help/glossary/zoom-level/) (since keplerGL uses Mapbox). On this basis the new strategy is:

```
1. Scan all data points and find max X, max Y, min X, min Y
2. Calculate midpointX ((maxX+minX)/2) and midpointY ((maxY+minY)/2)
3. Set viewport centre to midpointX, midpointY
4. Set viewport zoom level to 3 ( continent )
```

(from Mapbox docs:)
<img width="579" alt="Screenshot 2023-01-20 at 16 40 19" src="https://user-images.githubusercontent.com/1473244/213754171-3c7374ab-ef22-46ca-a996-f2517d36c152.png">

It's not ideal, because sometimes we may want city level. Will review this option with team.

**Comment by alexbfree** [comment]
UPDATE:

After further investigation of https://github.com/digipower-academy/private/issues/22, i can now say categorically that :

- There is no default centring or zoom within the kepler compoonent.
- Many of our experience tabs do default to Geneva - **but only if there is no locatable data** - ie. no lat/long columns, not necessarily no rows, on the file(s) that the map tab in question reads from.
- If there is geolocatable data in the data for the specific tab where the kepler map is, Kepler handles the centring of the map automatically base on the passed in data, and this is working correctly in all cases I have tested.

Therefore, there is nothing to do here and I will close this feature request.

The only open question is what to do about default behaviour when there is no geodata.

Options include:

a) hide the map and only show the table - because there is nothing to show
b) zoom to Geneva
c) zoom to whole continent of Europe (or some other place? whole world?)
d) look at data across all tabs, figure out centre & zoom base on that. 
e) have a configuration option per bubble/deployment which allows a default zoom location/zoom level to be specified
f) it doesn't matter, we don't even care about this scenario, now we know what is happening, we can just explain it's due to no geodata.
g) some other behaviour. 

This can be handled as part of ticket https://github.com/digipower-academy/private/issues/22

BTW, if we ever do need an improved zoom, I checked in some code on [default-zoom-enhancement](https://github.com/hestiaAI/hestialabs-experiences/tree/default-zoom-enhancement) branch that may be useful. But i think Kepler default behaviour is good enough and this branch can be ignored/archived/deleted.
